### PR TITLE
Add insert to Data.Fin.Permutations as rough inverse to remove

### PR DESCRIPTION
**Pull request by Taneb**
I defined this in the progress of proving that there's an isomorphism between `Fin (n !)` and `Pemutation m n` but I haven't finished that yet.

### CODE REVIEW

**Comment by JacquesCarette** [comment]
You are aware of [Symmetries in Reversible Programming: From Symmetric Rig Groupoids to Reversible Programming Languages](https://arxiv.org/abs/2110.05404) ? And the [artifacts](https://github.com/vikraman/popl22-symmetries-artifact)  that go along? 

That paper uses HoTT-Agda, so it's entirely possible that none of it can be easily re-used. But you're definitely playing in the same area.

**Comment by Taneb** [comment]
> You are aware of [Symmetries in Reversible Programming: From Symmetric Rig Groupoids to Reversible Programming Languages](https://arxiv.org/abs/2110.05404) ? And the [artifacts](https://github.com/vikraman/popl22-symmetries-artifact) that go along?
> 
> That paper uses HoTT-Agda, so it's entirely possible that none of it can be easily re-used. But you're definitely playing in the same area.

I hadn't seen that! Thanks for the link, I'll check it out.

File: src/Data/Fin/Permutation.agda [hunk 1]
```diff
@@ -246,3 +287,30 @@ lift₀-transpose i j (suc k) with does (k ≟ i)
...
+remove-insert : ∀ i j (π : Permutation m n) → remove i (insert i j π) ≈ π
+remove-insert i j π k with i ≟ i
+... | yes _ = begin
+  punchOut {i = j} _
+    ≡⟨ punchOut-cong j (insert-punchIn i j π k) ⟩
+  punchOut {i = j} (punchInᵢ≢i j (π ⟨$⟩ʳ k) ∘ sym)
+    ≡⟨ punchOut-punchIn j ⟩
+  π ⟨$⟩ʳ k
+    ∎
+... | no i≢i = ⊥-elim (i≢i refl)
```
Comment (MatthewDaggitt) [inline src/Data/Fin/Permutation.agda:316]: Can we use `contradiction` here instead? Same for other uses of `⊥-elim`. Also the proof would be slightly more readable if you put the `no` above the `yes` in my opinion.

File: src/Data/Fin/Permutation.agda [hunk 2]
```diff
@@ -246,3 +287,30 @@ lift₀-transpose i j (suc k) with does (k ≟ i)
 ... | false with does (k ≟ j)
 ...   | false = refl
 ...   | true = refl
+
+insert-punchIn : ∀ i j (π : Permutation m n) k → insert i j π ⟨$⟩ʳ punchIn i k ≡ punchIn j (π ⟨$⟩ʳ k)
+insert-punchIn i j π k with i ≟ punchIn i k
+... | yes i≡punchInᵢk = ⊥-elim (punchInᵢ≢i i k (sym i≡punchInᵢk))
+... | no  i≢punchInᵢk = begin
+  punchIn j (π ⟨$⟩ʳ punchOut i≢punchInᵢk) ≡⟨ cong (λ l → punchIn j (π ⟨$⟩ʳ l)) (punchOut-cong i refl) ⟩
```
Comment (MatthewDaggitt) [inline src/Data/Fin/Permutation.agda:295]: Can you align the eq reasoning operators in this and the proof below?

File: src/Data/Fin/Permutation.agda [hunk 3]
```diff
@@ -246,3 +287,30 @@ lift₀-transpose i j (suc k) with does (k ≟ i)
 ... | false with does (k ≟ j)
 ...   | false = refl
 ...   | true = refl
+
+insert-punchIn : ∀ i j (π : Permutation m n) k → insert i j π ⟨$⟩ʳ punchIn i k ≡ punchIn j (π ⟨$⟩ʳ k)
```
Comment (MatthewDaggitt) [inline src/Data/Fin/Permutation.agda:291]: This proof hasn't been added to the changelog?

File: src/Data/Fin/Permutation.agda [hunk 4]
```diff
@@ -189,6 +190,46 @@ lift₀ {m} {n} π = permutation to from record
...
+  right-inverse-of : ∀ k → to (from k) ≡ k
+  right-inverse-of k with j ≟ k
+  ... | yes j≡k rewrite proj₂ (dec-yes (i ≟ i) refl) = j≡k
+  ... | no  j≢k with dec-no (i ≟ punchIn i (π ⟨$⟩ˡ punchOut j≢k)) (punchInᵢ≢i i (π ⟨$⟩ˡ punchOut j≢k) ∘ sym)
+  ... | i≢punchInᵢπˡpunchOutj≢k , p rewrite p = begin
+    punchIn j (π ⟨$⟩ʳ punchOut i≢punchInᵢπˡpunchOutj≢k)                    ≡⟨ cong (λ l → punchIn j (π ⟨$⟩ʳ l)) (punchOut-cong i refl) ⟩
+    punchIn j (π ⟨$⟩ʳ punchOut (punchInᵢ≢i i (π ⟨$⟩ˡ punchOut j≢k) ∘ sym)) ≡⟨ cong (λ l → punchIn j (π ⟨$⟩ʳ l)) (punchOut-punchIn i) ⟩
+    punchIn j (π ⟨$⟩ʳ (π ⟨$⟩ˡ punchOut j≢k))                              ≡⟨ cong (punchIn j) (inverseʳ π) ⟩
+    punchIn j (punchOut j≢k)                                              ≡⟨ punchIn-punchOut j≢k ⟩
+    k ∎
```
Comment (MatthewDaggitt) [inline src/Data/Fin/Permutation.agda:231]: Alignment of `\qed`

**Pull request merged**